Build new careers meganav

carkod opened the pull request:
## Done

Build new careers navigation

## QA

- Check out this feature branch
- Run the site using the command `./run serve`
- View the site locally in your web browser at: http://0.0.0.0:8002/
- Run through the following [QA steps](https://discourse.canonical.com/t/qa-steps/152)
- [List additional steps to QA the new features or prove the bug has been resolved]

## Issue / Card

Fixes #

## Screenshots

[if relevant, include a screenshot]

webteam-app commented:
Demo starting at https://canonical-com-932.demos.haus

lyubomir-popov commented:
Hi @carkod, nice work, a few small  things:

- this transition affects the font-size change at 1681px, please target the properties you need only:
![image](https://github.com/canonical/canonical.com/assets/2741678/6620df86-c39c-4b9a-a7b7-b5d9c9e9af25)

- please use the link colours on dark (both default and visited) - they are in vanilla 4.0, which is already released, pls see the documentation). The current colours are not accessible forthe dark theme.

This is what it should look like: 
![image](https://github.com/canonical/canonical.com/assets/2741678/4c703024-a99d-4406-93f3-c1cc795313e7)


- on smaller breakpoints, the lines do not appear between all rows:
![image](https://github.com/canonical/canonical.com/assets/2741678/724f5be8-575b-441d-a0dc-48c512e19485)

- No need for break in the Careers overview on small screens:
![image](https://github.com/canonical/canonical.com/assets/2741678/867e5417-359d-4a07-97a7-d42a6a0c01d4)

- lines only between some items
![image](https://github.com/canonical/canonical.com/assets/2741678/2abff024-ecd5-4196-a896-c0bb47024b24)

carkod commented:
> Hi @carkod, nice work, a few small things:
> 
> * this transition affects the font-size change at 1681px, please target the properties you need only:
>   ![image](https://user-images.githubusercontent.com/2741678/251684256-6620df86-c39c-4b9a-a7b7-b5d9c9e9af25.png)
> * please use the link colours on dark (both default and visited) - they are in vanilla 4.0, which is already released, pls see the documentation). The current colours are not accessible forthe dark theme.
> 
> This is what it should look like: ![image](https://user-images.githubusercontent.com/2741678/251685154-4c703024-a99d-4406-93f3-c1cc795313e7.png)
> 
> * on smaller breakpoints, the lines do not appear between all rows:
>   ![image](https://user-images.githubusercontent.com/2741678/251685316-724f5be8-575b-441d-a0dc-48c512e19485.png)
> * No need for break in the Careers overview on small screens:
>   ![image](https://user-images.githubusercontent.com/2741678/251685424-867e5417-359d-4a07-97a7-d42a6a0c01d4.png)
> * lines only between some items
>   ![image](https://user-images.githubusercontent.com/2741678/251685492-2abff024-ecd5-4196-a896-c0bb47024b24.png)

Addressed some of these comments except:

- **No need for break in the Careers overview on small screens:** what do you mean? In the design the only place that this title is broken is in the small screens (xs)
- **this transition affects the font-size change at 1681px, please target the properties you need only**. I removed some of the uncessary styles, but most of the animations and styles come from u.com for consistency purposes.

lyubomir-popov commented:
The linksare misaligned: 
<img width="814" alt="image" src="https://github.com/canonical/ubuntu.com/assets/2741678/d741a17c-7110-4703-9229-9df609eba4f2">
the colour of links is still not the dark link colours recently introdued in vanilla:
![Uploading image.png…]()

lyubomir-popov commented:
Nice work @carkod the colours are correct now. I've given +1, optionally I'd get rid of the p-blocks around individual links as they make things quite spaced out on mobile. But other than that - looks great, thanks!

lyubomir-popov commented:
@anthonydillon I already reviewed, pointed out themust haves and the nice to haves here, I'm ok for this to be merged. If the optional requests above can't be done now, would be nice to have them as a follow up.

anthonydillon commented:
QA issues:
 - [ ] The Partners dropdown doesn't work on small screens
 - [ ] Its disorientating to jump down when clicking Partners on large screens
 - [ ] When tabbing I have to tab through the entire Careers menu even when it's hidden

anthonydillon commented on static/sass/_pattern_navigation.scss:
```diff
@@ -0,0 +1,112 @@
+    > li:not(:first-child)
+    > .p-navigation__dropdown-item::before {
+      @media (min-width: $breakpoint-navigation-threshold) {
+        background: rgba(255 255 255 / 10%);
```
I believe its `background-color: rgba(255, 255, 255, 0.1)`

anthonydillon commented on static/sass/_pattern_navigation.scss:
```diff
@@ -0,0 +1,112 @@
+}
+
+.dropdown-window {
+	@include vf-animation(all, brisk, in);
```
Tabbing is off here

anthonydillon commented on templates/partial/_careers_navigation.html:
```diff
@@ -0,0 +1,334 @@
+<div class="dropdown-window-overlay fade-animation"></div>
```
All this tabbing looks off

anthonydillon commented on static/sass/styles.scss:
```diff
@@ -485,12 +438,9 @@ $p-small-lh-diff: map-get($line-heights, default-text) - map-get($line-heights,
-
-  background-color: #f2f2f2;
+// Adjustment for career navigation
+.muted-color {
+  color: #a7a7a7;
 }
```
You don't need this if you use p-strip--dark

carkod replied:
`p-strip` adds additional spacing that will not fit with design's spacing requirements.

anthonydillon commented on templates/partial/_careers_navigation.html:
```diff
@@ -0,0 +1,334 @@
+		<div class="u-fixed-width p-strip">
+			<h2><a class="p-link is-dark" href="/careers">Careers <br class="u-hide--medium u-hide--large" />overview</a></h2>
+		</div>
+		<div class="row">
```
This row is without a strip

carkod replied:
That doesn't seem to work? p-strip still can't change the `span` color property

anthonydillon commented on templates/partial/_careers_navigation.html:
```diff
@@ -0,0 +1,334 @@
+<div class="dropdown-window slide-animation">
+	<div class="u-hide p-section" id="careers-menu" aria-hidden="true">
+		<div class="u-fixed-width p-strip">
+			<h2><a class="p-link is-dark" href="/careers">Careers <br class="u-hide--medium u-hide--large" />overview</a></h2>
```
I don't think p-link does anything

anthonydillon commented on templates/partial/_careers_navigation.html:
```diff
@@ -0,0 +1,334 @@
+<div class="dropdown-window-overlay fade-animation"></div>
+<div class="dropdown-window slide-animation">
+	<div class="u-hide p-section" id="careers-menu" aria-hidden="true">
+		<div class="u-fixed-width p-strip">
```
Fixed width should be within the strip. Strips add vertical padding

carkod replied:
I couldn't find a way to do it without breaking the design spacing.

anthonydillon commented on templates/partial/_careers_navigation.html:
```diff
@@ -0,0 +1,334 @@
+				<div class="u-sv2">
+					<h4 class="p-muted-heading p-text--small-caps muted-color">Roles by department</h4>
+				</div>
+				<hr class="p-rule is-dark is-bold">
```
`is-bold` doesn't do anything

carkod replied:
Removed

anthonydillon commented on templates/partial/_careers_navigation.html:
```diff
@@ -0,0 +1,334 @@
+				</div>
+				<hr class="p-rule is-dark is-bold">
+				<div class="row">
+					<div class="col-2 col-medium-2 col-small-12 p-block">
```
`p-block` doesn't do anything here

carkod replied:
It adds spacing on the bottom for smaller screens

anthonydillon commented on templates/partial/_careers_navigation.html:
```diff
@@ -0,0 +1,334 @@
+
+<script type="text/javascript">
+
+	var nav = document.querySelector('.js-show-nav');
+  var hash = window.location.hash;
+	var dropdownWindow = document.querySelector(".dropdown-window");
```
Tabbing

carkod replied:
Code moved to `navigation.js`

anthonydillon commented on templates/partial/_careers_navigation.html:
```diff
@@ -0,0 +1,334 @@
+  }
+
+	function keyPressHandler (e) {
+    console.log(e);
```
Console log

anthonydillon commented on templates/partial/_careers_navigation.html:
```diff
@@ -0,0 +1,334 @@
+	function keyPressHandler (e) {
+    console.log(e);
+    if (e.key === "Escape") {
+      closeSearch();
```
There is no closeSearch function

anthonydillon commented on templates/partial/_careers_navigation.html:
```diff
@@ -0,0 +1,334 @@
+      console.warn("Hash " + hash + " not found in topnav");
+    }
+    if (selected) {
+      selected.onmouseover();
```
This is not valid code

carkod replied:
Removed

anthonydillon reviewed (dismissed):
Some QA issues and quite a few comments in the code.

carkod commented on static/sass/_pattern_navigation.scss:
```diff
@@ -0,0 +1,112 @@
+    position: relative;
+  }
+  
+  .p-navigation__dropdown
```
Removed

mtruj013 commented on static/sass/_pattern_navigation.scss:
```diff
@@ -0,0 +1,70 @@
+	width: 100%;
+	z-index: 3;
+
+	&.slide-animation {
+		box-shadow: none;
+	}
+
+	&__content {
+		background-color: $color-dark;
+		color: $color-dark;
+	}
+
+	& .p-inline-list__item {
+		list-style-type: none;
+
+		&::after {
+			@media only screen and (max-width: $breakpoint-small) {
+				display: none;
+			}
+		}
+	}
```
Looks like the indentation here needs fixing

mtruj013 commented on templates/partial/_careers_navigation.html:
```diff
@@ -0,0 +1,190 @@
+<div class="dropdown-window slide-animation" aria-hidden="true" aria-expanded="false">
```
Could you fix the indentation on this file? I believe our convention is 2 spaces

mtruj013 commented on templates/partial/_navigation.html:
```diff
@@ -37,6 +37,7 @@
         </li>
         <li class="p-navigation__item" id="products-nav">
           <a href="/#products" class="p-navigation__link">Products</a>
+          
```
```suggestion
```

mtruj013 reviewed (approved):
Thanks @carkod, looks great! Just some minor comments in the code. Also I think you might need to rebase, not sure why but all the dept pages with the exception of engineering are showing up with the old styles both in the demo and on local

lyubomir-popov commented:
Two more comments from me:
- can we please make the animation come to a smooth stop, rather than accellerating smoothly, then coming to an abrupt stop? (Ease out not ease in)
- can we please make the lines muted? (add class `is-muted`): there isn't much content in there, so we don't need such strong separation. It will look like so: 
<img width="1577" alt="image" src="https://github.com/canonical/canonical.com/assets/2741678/3d1027ba-cda0-4586-986f-51780c8bbc93">

carkod commented on static/js/navigation.js:
```diff
@@ -130,3 +143,115 @@ function addGAContentEvents(target) {
+            `[data-slug='${dpt.slug}']`
+          );
+          departmentElement.innerText = `${dpt.count} roles`;
+          departmentElement.classList.add("u-animation--slide-from-top");
```
This is used in the careers progression page. So here I'm just reusing it.

carkod commented on static/sass/_pattern_navigation.scss:
```diff
@@ -0,0 +1,70 @@
+}
+
+// Meganav
+.slide-animation {
```
This is probably the best thing I've done in this feature. It does this slight animation, which makes it look cool, but it's actually because we are making a call to greenhouse, and it goes down quite frequently according to sentry, so making a JS request should protect the site from going down (navigation is loaded everywhere)

carkod commented:
> Two more comments from me:
> 
> * can we please make the animation come to a smooth stop, rather than accellerating smoothly, then coming to an abrupt stop? (Ease out not ease in)
> * can we please make the lines muted? (add class `is-muted`): there isn't much content in there, so we don't need such strong separation. It will look like so:
> 

Done. I was able to tweak the Vanilla mixing to do `ease out`. Cool.
